hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -736,23 +744,48 @@ public void notifyTermIndexUpdated(long term, long index) {
     return f;
   }
 
+  // Removes the stateMachine data from cache once both followers catch up
+  // to the particular index.
+  private void removeStateMachineDataIfNeeded(long index) {
+    if (waitOnBothFollowers) {
+      LOG.info("Removing data corresponding to log index {} from cache",
+          index);
+      try {
+        RaftServer.Division division = ratisServer.getServer().getDivision(gid);
+        if (division.getInfo().isLeader() && Arrays
+            .stream(division.getInfo().getFollowerNextIndices())
+            .allMatch(i -> i >= index)) {
```

Comment:
The probability of this being true is still less. How about we take a min of next indexes and index and then remove the minimum from the statemachine cache?

Reply:
The idea is to wait for both followers to catch up . How minimum is useful here? 
As soon as both followers catch up, the data will be cleaned up on next index update/applyTransaction.

There is already an alternative approach addressed with https://issues.apache.org/jira/browse/RATIS-1441 which changes the majority index based on the min index and gap threshold. 

The idea of this patch is to have all solutions built and then tried out in a real cluster and see what works best.

Reply:
The comment is for slow follower case. In case of a slow follower `allMatch(i -> i >= index)` would always fail and as a result eviction would never happen.

Reply:
In case of a slow follower, stateMachine cache will be full and then will block the client. This is what is expected with this change. After that, the pipeline will be teared down. The idea is to make sure the slow follower catch up and if that doesn't work out, stall the pipeline and close the pipeline. Client will also exclude the slow datanodes.

Reply:
It is a slow follower but still accepting requests. In that case next index for one follower can be lagging but it will catch up eventually. 
```
        if (division.getInfo().isLeader() && Arrays
            .stream(division.getInfo().getFollowerNextIndices())
            .allMatch(i -> i >= index)) {
```
But this condition will never be satisfied. Next index for one follower is always less than index. Index itself is increasing.  As a result eviction would never be triggered.

Reply:
Agreed. Addressed in the latest patch.